rtos/ConditionVariable.cpp

```diff
@@ -0,0 +1,128 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2012 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "ConditionVariable.h"
+
+#include "mbed_error.h"
+#include "mbed_assert.h"
+
+namespace rtos {
+
+#define RESUME_SIGNAL      (1 << 15)
+
+class Waiter {
+public:
+    Waiter();
+    osThreadId thread;
+    Waiter *prev;
+    Waiter *next;
+    bool in_list;
+};
+
+Waiter::Waiter(): thread(osThreadGetId()), prev(NULL), next(NULL), in_list(false) {
+    // No initialization to do
+}
+
+ConditionVariable::ConditionVariable(Mutex &mutex): _mutex(mutex), wait_list(NULL) {
+    // No initialization to do
+}
+
+bool ConditionVariable::wait(uint32_t millisec) {
+    Waiter current_thread;
+    MBED_ASSERT(_mutex.is_owner());
+    add_wait_list(&current_thread);
+
+    // Unlock mutex
+    uint32_t count = _mutex.count;
+    for (uint32_t i = 0; i < count; i++) {
+        _mutex.unlock();
+    }
+
+    osEvent event = osSignalWait(RESUME_SIGNAL, millisec);
+    bool ret = (osEventSignal == event.status) ? true : false;
+
+    // Lock mutexs
+    for (uint32_t i = 0; i < count; i++) {
+        _mutex.lock();
+    }
+
+    if (current_thread.in_list) {
+        remove_wait_list(&current_thread);
+    }
+
+    return ret;
+}
+
+void ConditionVariable::notify() {
+    MBED_ASSERT(_mutex.is_owner());
```

Comment:
Could do, but would make it harder to debug. I strongly favour asserts for programming errors, rather than the two bad choices of adding run-time code to catch and report a returned error that should never happen, or not bother to check for errors.

For comparison - do you think memcpy should return an error code for bad pointers, and people should check for the error return on every memcpy?